patches/api/0449-add-number-format-api.patch

```diff
@@ -0,0 +1,194 @@
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/scoreboard/numbers/FixedFormatImpl.java
+@@ -0,0 +1,10 @@
++package io.papermc.paper.scoreboard.numbers;
++
++import net.kyori.adventure.text.Component;
++
++record FixedFormatImpl(Component component) implements FixedFormat {
++    @Override
++    public Component getComponent() {
++        return component;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/scoreboard/numbers/NumberFormat.java b/src/main/java/io/papermc/paper/scoreboard/numbers/NumberFormat.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e326afc55cb0a3519471cbfc5ea44519f19b7a9c
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/scoreboard/numbers/NumberFormat.java
+@@ -0,0 +1,40 @@
++package io.papermc.paper.scoreboard.numbers;
++
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.format.Style;
++import org.jetbrains.annotations.NotNull;
++
++/*
++ * Describes a scoreboard number format that applies custom formatting to scoreboard scores.
++ */
++public interface NumberFormat {
++
++    /*
++     * Creates a blank scoreboard number format that removes the score number entirely.
++     *
++     * @return a blank number format
++     */
++    static NumberFormat blank() {
++        return BlankFormatImpl.INSTANCE;
++    }
++
++    /*
++     * Creates a scoreboard number format that applies a custom formatting to the score number.
++     *
++     * @param style the style to apply on the number
++     * @return a styled number format
++     */
++    static StyledFormat styled(@NotNull Style style) {
++        return new StyledFormatImpl(style);
++    }
++
++    /*
++     * Creates a scoreboard number format that replaces the score number with a chat component.
++     *
++     * @param component the component to replace the number with
++     * @return a fixed number format
++     */
++    static FixedFormat fixed(@NotNull Component component) {
++        return new FixedFormatImpl(component);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d09726c6821a9463682cf65bde21f25932e2856f
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java
+@@ -0,0 +1,23 @@
++package io.papermc.paper.scoreboard.numbers;
++
++import net.kyori.adventure.text.format.NamedTextColor;
++import net.kyori.adventure.text.format.Style;
++
++/*
++ * A scoreboard number format that applies a custom formatting to the score number.
++ */
++public interface StyledFormat extends NumberFormat {
++
++    StyledFormat NO_STYLE = NumberFormat.styled(Style.empty());
++
++    StyledFormat SIDEBAR_DEFAULT = NumberFormat.styled(Style.empty().color(NamedTextColor.RED));
++
++    StyledFormat PLAYER_LIST_DEFAULT = NumberFormat.styled(Style.empty().color(NamedTextColor.YELLOW));
```

Comment:
I think we should make these static methods that return these constants that can then be declared inside the package private record. similar to the blank instance.

Reply:
I'm kind of against that. Having it like that shows the implementation directly to the devs that use this api. 

Although these constants might change in order to stay consistent with Mojangs implementation, I don't think they ever will. So I'd prefer people to see what it actually does

Reply:
They can still see them if they goto the source file. Package-private api source is still visible in the IDE or whatever, there isn't an issue with visibility. I think it is just cleaner if its in the impl instead of the interface. 

Incidentally that's another reason for the interface+record pattern is keeping the interface as straightforward as possible.

Reply:
Well, you're right, but I just think it's a bit easier to follow along when it's in the interface itself

Reply:
If those values ever end up depending on internals, it is a lot easier for us to change things around if the values are exposed via methods vs fields.

Grabbing a different method result (via e.g. ServiceProvider) is a lot nicer than redefining these fields to a result of a ServiceProvider provided instance's method (e.g. `ServiceProvider.get(ScoreboardStyles.class).blablabla()` )

Reply:
Should I add these methods to the StyledFormat interface, or would you guys prefer to have that in the number format interface along the other methods?

Reply:
> Should I add these methods to the StyledFormat interface, or would you guys prefer to have that in the number format interface along the other methods?

@Machine-Maker

Reply:
I added them to the styled format interface for now. Let me know if I should change it